lib/timer.py

```diff
@@ -16,68 +15,72 @@
-            "job_func": update_apt_and_motd_messages,
-            "run_interval_seconds": 43200,  # 12 hours
-        }
-    }
-)  # type: Dict[str, Dict[str, Any]]
-
-
-def run_job(
-    job_name: str, job_func: "Callable[[UAConfig], None]", cfg: UAConfig
-) -> bool:
-    """Run a job in a failsafe manner, returning True on success.
-
-    :param job_name: Name of the job to run
-    :param job_func: Job callable to call
-    :param cfg: UAConfig instance
-
-    :return: A bool True when successfully run. False if ignored or in error.
-    """
-    LOG.debug("Running job: %s", job_name)
-    try:
-        job_func(cfg)
-    except Exception as e:
-        LOG.warning("Error executing job %s: %s", job_name, str(e))
-        return False
-
-    return True
-
-
-def run_jobs(cfg: UAConfig, current_time: datetime):
-    """Run ordered UACLIENT_JOBS when next_run is before utcnow.
-
-    Persist jobs-status with calculated next_run values to aid in timer state
-    introspection for jobs which ave not yet run.
-    """
-    LOG.debug("Trigger UA Timer jobs")
-    jobs_status = cfg.read_cache("jobs-status") or {}
-    for job_name, job_info in UACLIENT_JOBS.items():
-        if job_name in jobs_status:
-            next_run = datetime.strptime(
-                jobs_status[job_name]["next_run"], "%Y-%m-%dT%H:%M:%S.%f"
-            )
-            if next_run > current_time:
-                continue  # Skip job as expected next_run hasn't yet passed
-        job_func = job_info["job_func"]  # type: Callable[[UAConfig], None]
-        if run_job(job_name=job_name, job_func=job_func, cfg=cfg):
-            # Persist last_run and next_run UTC-based times on job success.
-            jobs_status[job_name] = {
-                "last_run": current_time,
-                "next_run": current_time
-                + timedelta(seconds=job_info["run_interval_seconds"]),
-            }
-    cfg.write_cache(key="jobs-status", content=jobs_status)
+
+class TimedJob:
+    def __init__(
+        self, name: str, job_func: Callable, default_interval_seconds: int
+    ):
+        self.name = name
+        self._job_func = job_func
+        self._default_interval_seconds = default_interval_seconds
+
+    def run(self, cfg: UAConfig):
+        """Run a job in a failsafe manner, returning True on success.
+
+        :param cfg: UAConfig instance
+
+        :return: A bool True when successfully run. False if ignored or in error.
+        """
+        LOG.debug("Running job: %s", self.name)
+        try:
+            self._job_func(cfg)
+        except Exception as e:
+            LOG.warning("Error executing job %s: %s", self.name, str(e))
+            return False
+
+        return True
+
+    def run_interval_seconds(self, cfg: UAConfig):
+        """Return the run_interval for the job based on config or defaults"""
+        return cfg.timer.get(self.name, {}).get(
+            "run_interval_seconds", self._default_interval_seconds
+        )
+
+
+class Timer:
```

Comment:
Why do we need a Timer class ?

Reply:
We do not - I thought it would be a more organized and readable way to encapsulate the jobs and the running functionality. and since Jobs were turned into classes, I created the Timer one as well.

Reply:
I like have you've ecapsulated the TimedJob.run definition, but I'm a little uncertain if this is overengineering the "Timer" class at the moment since we are only really talking about an ordered dict and a single function to run_jobs. I also have a general aversion to __init__ methods in classes which have a lot of side-effects (like initializing UAConfig and multiple TimedJob instances. Are there other compelling reasons to make this a class?

Reply:
No other compelling. Will change it back to a function.